[train][doc] Add validation and other details to checkpoint doc (#57065)
ray-project/ray: doc/source/train/user-guides/checkpoints.rst

.. literalinclude:: ../doc_code/checkpoints.py
    :language: python
    :start-after: __checkpoint_upload_mode_sync_start__
    :end-before: __checkpoint_upload_mode_sync_end__

Asynchronous checkpoint uploading
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You may want to upload your checkpoint asynchronously instead so that
the next training step can start in parallel. If so, you should use
``ray.train.CheckpointUploadMode.ASYNC``, which kicks off a new thread
to upload the checkpoint. This is helpful for larger
checkpoints that might take longer to upload, but might add unnecessary
complexity (see below) if you want to immediately upload only a small checkpoint.

Each ``report`` blocks until the previous ``report``\'s checkpoint
upload completes before starting a new checkpoint upload thread. Ray Train does this
to avoid accumulating too many upload threads and potentially running out of memory.

Because ``report`` returns without waiting for the checkpoint upload to complete,
you must ensure that the local checkpoint directory stays alive until the checkpoint
upload completes. This means you cannot ``report`` a checkpoint whose directory is

> from a ``tempfile.TemporaryDirectory`` context manager. ``report`` also exposes the

Author: done - new description is more general indeed.

``ray.train.CheckpointUploadMode.ASYNC``, which kicks off a new thread
to upload the checkpoint. This is helpful for larger
checkpoints that might take longer to upload, but might add unnecessary
complexity (see below) if you want to immediately upload only a small checkpoint.

Each ``report`` blocks until the previous ``report``\'s checkpoint
upload completes before starting a new checkpoint upload thread. Ray Train does this
to avoid accumulating too many upload threads and potentially running out of memory.

Because ``report`` returns without waiting for the checkpoint upload to complete,
you must ensure that the local checkpoint directory stays alive until the checkpoint
upload completes. This means you cannot use a temporary directory that may be

> deleted before the upload finishes (e.g. from ``tempfile.TemporaryDirectory``).

Reviewer: ```suggestion
delete before the upload finishes, for example from ``tempfile.TemporaryDirectory``.
```

* Running validation in parallel with the training loop
* Running validation on different hardware than training
* Leveraging :ref:`vms-autoscaling` to only use validation resources when needed

.. _train-validate-fn:

Writing your own validation function
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here is an example ``validate_fn`` that uses a :class:`~ray.train.torch.TorchTrainer`
to calculate average cross entropy loss on a validation set. Note that this ``report``s
a dummy checkpoint so that the ``TorchTrainer`` keeps the metrics.

.. literalinclude:: ../doc_code/checkpoints.py
    :language: python
    :start-after: __validation_fn_torch_trainer_start__
    :end-before: __validation_fn_torch_trainer_end__

Here is an example ``validate_fn`` that uses :func:`~ray.data.Dataset.map_batches` to
calculate average cross entropy loss on a validation set. To learn more about how to use
``map_batches`` for batch inference, check out :ref:`batch_inference_home`.

.. literalinclude:: ../doc_code/checkpoints.py
    :language: python
    :start-after: __validation_fn_map_batches_start__
    :end-before: __validation_fn_map_batches_end__

You should use ``TorchTrainer`` if:

> * You don't want to add a dependency on Ray Data

Author: they could have used the torch dataloader instead but i think adding an additional case that does htis i smore confusing

    please ensure that the metric is always reported together with the checkpoints.

.. _train-validating-checkpoints:

Validating checkpoints asynchronously
-------------------------------------

You can also asynchronously validate checkpoints that you :func:`~ray.train.report` as follows:

* Define your own ``validation_fn`` whose inputs are the :class:`~ray.train.Checkpoint` to validate

>   and an optional ``validate_config`` dict and whose output is a dict of metrics from the validation. We recommend

Author: done.

>   and an optional ``validate_config`` dict and whose output is a dict of metrics from the validation. We recommend

Author: done.

    please ensure that the metric is always reported together with the checkpoints.

.. _train-validating-checkpoints:

> Validating checkpoints asynchronously

Reviewer 2: Recommend reordering this so it's easier to follow conceptually.

1. Introduce the problem (the benefits you've listed)
2. Outline the steps.
     a. Show a simple example here (with just a single model, non-distributed).
     b. The main goal is so that the user understands how `validate_fn`, `validate_config`, `report`, `checkpoint` integrate in their existing training function.
3. Solutions for running distributed (Train, Data)

-------------------------------------

You can also asynchronously validate checkpoints that you :func:`~ray.train.report` as follows:

* Define your own ``validate_fn`` that takes a :class:`~ray.train.Checkpoint` to validate
  and an optional ``validate_config`` dictionary, which contain arguments needed for validation
  such as the validation dataset, and returns a dictionary of metrics from the validation.
  We recommend performing the validation with an eval-only :ref:`Trainer <train-overview-trainers>` or with
  :func:`~ray.data.Dataset.map_batches` - see :ref:`train-validate-fn` for more details.
* Call :func:`~ray.train.report` with your ``validate_fn`` and a ``validate_config`` dict.
  Ray Train runs your ``validate_fn`` with the ``validation_config`` and ``checkpoint``
  in a new Ray task. When that task completes, Ray Train associates the metrics returned by

>   ``validate_fn`` with that ``checkpoint``.

Reviewer 2: This is a bit heavy on the text. Consider nested bullet points, individual headers, or putting more information in the API docstrings themselves.

The main benefits of validating with :func:`~ray.train.report` over performing the validation
in the training loop include:

* Running validation in parallel with the training loop
* Running validation on different hardware than training

> * Leveraging :ref:`vms-autoscaling` to rent user-specified machines only for the duration of the validation

Reviewer 2: nit: Rename this so it reads better.

```suggestion
* Leveraging :ref:`Autoscaling <vms-autoscaling>` to rent user-specified machines only for the duration of the validation
```

<img width="836" height="44" alt="Image" src="https://github.com/user-attachments/assets/8d33edfd-897c-4b6e-bdcd-9a7b45b29c4c" />

Here is an example ``validate_fn`` that uses a :class:`~ray.train.torch.TorchTrainer`
to calculate average cross entropy loss on a validation set. Note that this ``report``\s
a dummy checkpoint so that the ``TorchTrainer`` keeps the metrics. Also note that while
the ``TorchTrainer`` is typically used for training, it can also be used for isolated
validation, and if you are using Anyscale, all these validation-only train runs will
show up on the `Train dashboard <https://docs.anyscale.com/monitoring/workload-debugging/train-dashboard>`.

.. literalinclude:: ../doc_code/checkpoints.py
    :language: python
    :start-after: __validate_fn_torch_trainer_start__
    :end-before: __validate_fn_torch_trainer_end__

Here is an example ``validate_fn`` that uses :func:`~ray.data.Dataset.map_batches` to
calculate average cross entropy loss on a validation set. To learn more about how to use
``map_batches`` for batch inference, check out :ref:`batch_inference_home`.

.. literalinclude:: ../doc_code/checkpoints.py
    :language: python
    :start-after: __validate_fn_map_batches_start__
    :end-before: __validate_fn_map_batches_end__

You should use ``TorchTrainer`` if:

> * You don't want to add a dependency on Ray Data

Reviewer 3: ```suggestion
* You want to keep your existing validation logic and avoid migrating to Ray Data. The training function API lets you fully customize the validation loop to match your current setup.
```

> * You prefer PyTorch's metrics aggregation API to Ray Data's

Reviewer 3: ```suggestion
* Your validation code depends on running within a Torch process group—for example, your metric aggregation logic uses collective communication calls, or your model parallelism setup requires cross-GPU communication during the forward pass.
```

You should use ``map_batches`` if:

* You care about validation performance. Preliminary benchmarks show that ``map_batches`` is
  faster and the Ray Data team will continue to work on OSS and proprietary performance improvements.

> * You prefer Ray Data's metrics aggregation API to PyTorch's

Reviewer 3: ```suggestion
* You prefer Ray Data’s native metric aggregation APIs over PyTorch, where you must implement aggregation manually using low-level collective operations or rely on third-party libraries such as `torchmetrics`.
```

Reporting with your validation function
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Once you've defined your ``validate_fn``, you can ``report`` it along with a ``validation_config``,

Reviewer 3: Can you highlight the fact that you shouldn't pass the validation dataset into the original Trainer's "datasets" arg?

Reviewer 3: Also, we should warn the user that they shouldn't capture large objects in this validate_fn, since it's run as a ray task and will serialize all captured variables.

That's why they need to package heavy things in the Checkpoint and access from shared storage later.
